### PR TITLE
Allow `inference.satisfiable` to handle inequalities by adding z3 as an optional dependency

### PR DESCRIPTION
**Pull request by TiloRC**
Enables `inference.satisfiable` to handle linear and polynomial inequalities. Currently still has some bugs.

#25363 will also add inequality handling. However, what sets this PR apart is that Z3 can handle polynomial inequalities while 25363 will only be able to handle linear inequalities. I thought Z3 would probably also be faster, but after running tests this seems to be very wrong.

Speed tests on `test_bounded()` from `test_query.py`. The ranges are rough estimates I made based on running the code 2 or 3 times.

| Satisfiable Algorithm | Time |
| --- | ----------- |
| dpll2 | 4.2 +/- .3 sec |
| z3 | 1 m 16 +/- 5 sec |

This time can't be improved upon very much because it mostly seems to be a slow down in Z3's python bindings or Z3 itself rather than my string parsing code. Perhaps there's some way to frame the problem differently so Z3 has an easier time with it.

It's possible for z3 to be unsure if something is sat or unsat — in such cases None is returned.

#### Release Notes

* logic
  * Added Z3 Theorem Prover as an optional algorithm for `satisfiable`. Can be used to handle linear and nonlinear inequalities.

### CODE REVIEW

File: sympy/logic/algorithms/z3_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from sympy.core import Add, Mul
+from sympy.core.relational import Equality, LessThan, GreaterThan, StrictLessThan, StrictGreaterThan
+from sympy.functions.elementary.complexes import Abs
+from sympy.functions.elementary.exponential import Pow
+from sympy.functions.elementary.miscellaneous import Min, Max
+from sympy.logic.boolalg import And, Or, Xor, Implies
+from sympy.logic.boolalg import Not, ITE
+from sympy.assumptions.assume import AppliedPredicate
+from sympy.assumptions.ask import Q
+from sympy.assumptions.relation.equality import StrictGreaterThanPredicate, StrictLessThanPredicate, GreaterThanPredicate, LessThanPredicate, EqualityPredicate
```
Comment (asmeurer) [inline sympy/logic/algorithms/z3_wrapper.py:97]: Move this to the top of the file.

**Comment by asmeurer** [comment]
We could test z3 in ask using something similar to https://github.com/sympy/sympy/blob/master/sympy/core/parameters.py

**Comment by asmeurer** [comment]
And https://github.com/sympy/sympy/blob/master/sympy/testing/pytest.py#L291

File: sympy/logic/inference.py [hunk 2]
```diff
@@ -105,6 +110,14 @@ def satisfiable(expr, algorithm=None, all_models=False, minimal=False):
     elif algorithm == "minisat22":
         from sympy.logic.algorithms.minisat22_wrapper import minisat22_satisfiable
         return minisat22_satisfiable(expr, all_models, minimal)
+    elif algorithm == "z3":
+        from sympy.logic.algorithms.z3_wrapper import z3_satisfiable
+        return z3_satisfiable(expr, all_models)
+
```
Comment (asmeurer) [inline sympy/logic/inference.py:116]: There should only be one blank line here

File: sympy/logic/tests/test_inference.py [hunk 3]
```diff
@@ -313,3 +313,15 @@ def test_satisfiable_all_models():
     result = satisfiable(Or(*X), all_models=True)
     for i in range(10):
         assert next(result)
+
+def test_z3():
```
Comment (asmeurer) [inline sympy/logic/tests/test_inference.py:317]: This should be skipped when z3 is not installed. See sympy/utilities/tests/test_lambdify.py for an example of how to do this.

File: sympy/assumptions/satask.py [hunk 4]
```diff
@@ -88,6 +88,9 @@ def check_satisfiability(prop, _prop, factbase):
     can_be_true = satisfiable(sat_true)
     can_be_false = satisfiable(sat_false)
 
+    if can_be_true is None or can_be_false is None:
+        return None
+
```
Comment (TiloRC) [inline sympy/assumptions/satask.py:93]: delete these changes

**Comment by asmeurer** [comment]
Make sure z3 is installed in the optional dependencies build on CI.

**Comment by TiloRC** [comment]
@asmeurer how do I make sure z3 is installed on the CI build? You explained it to me last week, but I forget how. 

sympy/logic is already in test_optional_dependencies.py.

File: .github/workflows/runtests.yml [hunk 5]
```diff
@@ -201,8 +201,11 @@ jobs:
       - run: python -m pip install --upgrade pip wheel setuptools
 
       # dependencies to install in all Python versions:
+              aesara wurlitzer autowrap lxml pytest z3                \
+
       - run: pip install numpy numexpr matplotlib ipython cython scipy aesara \
-                         wurlitzer autowrap lxml lark                         \
+                         wurlitzer autowrap lxml lark z3                       \
+
```
Comment (asmeurer) [inline .github/workflows/runtests.yml:208]: I think this blank line is breaking the YAML file, which is why the tests aren't running.

File: .github/workflows/runtests.yml [hunk 6]
```diff
@@ -201,8 +201,11 @@ jobs:
       - run: python -m pip install --upgrade pip wheel setuptools
 
       # dependencies to install in all Python versions:
+              aesara wurlitzer autowrap lxml pytest z3                \
```
Comment (asmeurer) [inline .github/workflows/runtests.yml:204]: Actually it's more likely this line, but the below blank line shouldn't be there either. The diff of this file should just be to add `z3` and nothing else.

**Comment by haru-44** [comment]
`sympy.logic.inference.satisfiable` is a function that returns a satisfiable variable assignment, correct?

```python
>>> from sympy.logic import satisfiable
>>> from sympy import symbols

>>> x, y, z = symbols('x y z')
>>> satisfiable(x & y | z)
{z: True, x: True, y: True}
```

I don't think that determining satisfiability would be in line with the other algorithms.

```python
>>> satisfiable(x & y | z, algorithm="z3")
True
```

**Comment by oscarbenjamin** [comment]
Maybe this should be using `.model()` rather than `.check()` or otherwise it should use the `solve` function:
```python
In [12]: from z3 import *

In [13]: x = Bool('x')

In [14]: y = Bool('y')

In [15]: z = Bool('z')

In [16]: s = Solver()

In [17]: s.add(x & y | z)

In [18]: s
Out[18]: [Or(And(x, y), z)]

In [19]: s.check()
Out[19]: sat

In [20]: s.model()
Out[20]: [z = False, y = True, x = True]

In [21]: solve(x & y | z)
[z = False, y = True, x = True]
```

**Comment by asmeurer** [comment]
Yes, it looks like this should be fixed. I wonder if Z3 (or other solvers) are somehow faster to return True than a full model. If so, it would be useful to have an API for that, since most uses of satisfiable don't actually use the model. If that's not the case, though, then we shouldn't worry about it.

**Comment by TiloRC** [comment]
> Maybe this should be using .model() rather than .check() or otherwise it should use the solve function

One problem with using `.model()` is that sometimes z3 doesn't provide a model. I'm not sure exactly when it does or when it doesn't. If I remember correctly, `.model()` doesn't work when using `.from_string` to add queries to the z3.Solver() object.

**Comment by oscarbenjamin** [comment]
This needs to be fixed somehow before the next release so I'm going to open a new issue,

**Pull request merged**